### ISSUE TITLE
Bug: [no-misused-promises] `flagUnions: 'strict'` misses unions with multiple Promise constituents

### ISSUE THREAD

**Issue by cchaeyoung** [typescript-eslint/typescript-eslint#12903]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.2&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3gNyhAriAFzwr4C2ARiDPAD7yYwCWKA5vfAAoxLnNoQAHjJUaAPk48%2BA4U1ZtxAbgCwAKGYAzeAApcBEAEp4Ab3UBfdUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMAbXDYciaOOiQANPIVYF2SGQAWiMgGtkAYQ5MhlfJUsBDeNMyadOAGbx7AcwCqTOyZnSGR8KjJ8SFcFAF9osDiFAF15OJigA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkAhqqrAO4BSZlemArugDTgakEl2TQ8AOW6MAwgAt00ANYcwXXoIC%2BITUA&tokens=false

### Repro Code

```TypeScript
declare const value: number | string | Promise<number> | Promise<string>;

if (value) {
}
```

### ESLint Config

```javascript
module.exports = {
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: true,
  },
  rules: {
    '@typescript-eslint/no-misused-promises': [
      'error',
      {
        checksConditionals: {
          flagUnions: 'strict',
        },
      },
    ],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {}
}
```

### Expected Result

The rule should report `if (value)` with `Expected non-Promise value in a boolean conditional.`, as it already does for `number | Promise<number>`.

With `flagUnions: "strict"`, the same should apply to `number | string | Promise<number> | Promise<string>`, since the awaited types of the Promise constituents exactly match the non-Promise constituents.

### Actual Result

No error is reported for `if (value)` when `value` has the type `number | string | Promise<number> | Promise<string>`.

In contrast, changing the type to `number | Promise<number>` while keeping the same `flagUnions: "strict"` setup correctly reports `Expected non-Promise value in a boolean conditional.`

### Additional Info

`hasMatchingPromiseTypeArgument` uses `.find()` to select a single Promise-like constituent.

When the union contains multiple Promise constituents, the remaining ones are treated as non-Promise constituents, causing the comparison to return `false`.

💖
